src/cluster.ml

```diff
@@ -0,0 +1,213 @@
+  | `Contents x -> Current.map (fun x -> `Contents x) x
+
+let component_label label dockerfile pool =
+  let pp_label = Fmt.(option (cut ++ string)) in
+  match dockerfile with
+  | `Path path -> Current.component "build %s@,%s%a" path pool pp_label label
+  | `Contents _ -> Current.component "build@,%s%a" pool pp_label label
+
+let ocluster_build ?level ?label ?cache_hint t ~pool ~src ~options dockerfile =
+  component_label label dockerfile pool |>
+  let> dockerfile = unwrap dockerfile
+  and> options
+  and> src in
+  Current_ocluster.Raw.build ?level ?cache_hint t ~pool ~src ~options dockerfile
+
+(* Build [src/dockerfile] as a Docker service. *)
+let build { sched; dockerfile; options; archs } ?(additional_build_args=Current.return []) repo src =
+  let options =
+    let+ additional_build_args = additional_build_args in
+    { options with build_args = additional_build_args @ options.build_args }
+  in
+  let hash = Current.map Current_git.Commit_id.hash src in
+  let build_arch arch =
+    let src = Current.map (fun x -> [x]) src in
+    let pool = Arch.pool_id arch in
+    let build = ocluster_build sched ~options ~pool ~src dockerfile in
+    let index =
+      let+ job_id = get_job_id build
+      and+ hash in
+      let label =
+        match dockerfile with
+        | `Path path -> Fmt.str "build %s@,%s" path pool
+        | `Contents _ -> Fmt.str "build@,%s" pool
+      in
+      Index.record ~repo ~hash [(label, job_id)]
+    in
+    Current.all [build; index]
+  in
+  Current.all (List.map build_arch archs)
+
+let name info = Cluster_api.Docker.Image_id.to_string info.hub_id
+
+let no_schedule = Current_cache.Schedule.v ()
+
+let pull_and_serve (module D : Current_docker.S.DOCKER) ~name op repo_id =
+  let image =
+    Current.component "pull" |>
+    let> repo_id in
+    Current_docker.Raw.pull repo_id ?auth ~docker_context:D.docker_context ~schedule:no_schedule
+    |> Current.Primitive.map_result (Result.map (fun raw_image ->
+        D.Image.of_hash (Current_docker.Raw.Image.hash raw_image)
+      ))
+  in
+  match op with
+  | `Service -> D.service ~name ~image ()
+  | `Compose contents ->
+      let contents = Current.map (fun image ->
+        Caddy.replace_hash_var ~hash:(D.Image.hash image) contents) image in
+      D.compose_cli ~name ~contents ~detach:true ()
+  | `Compose_cli contents ->
+      let contents = Current.map (fun hash ->
+        Aws.replace_hash_var ~hash contents) repo_id in
+      D.compose_cli ~name ~contents ~detach:false ()
+
+let build_and_push ?level ?label ?cache_hint t ~push_target ~pool ~src ~options dockerfile =
+  component_label label dockerfile pool |>
+  let> dockerfile = unwrap dockerfile
+  and> options
+  and> src in
+  Current_ocluster.Raw.build_and_push ?level ?cache_hint t ~push_target ~pool ~src ~options dockerfile
+
+let pull_and_serve multi_hash hub_id = function
+  (* deploy.ci.dev *)
+  | `Ci3 name -> pull_and_serve (module Ci3_docker) ~name `Service multi_hash
```

Comment:
If we decided to do away with the (afaict) unhelpful aliasing above, this could just be

```suggestion
  | `Ci3 name -> pull_and_serve (module Current_docker.Default) ~name `Service multi_hash
```

And then the same thin on line 178.

I think my rationale here is something like: providing two different names to the same value is generally not useful, as it obscures an identity. But when it is useful, it is because it helps document or clarify something. However, IMO, it is more helpful to read this line as "for service `Ci3, pull and serve the default docker module" rather than  "for service `Ci3 is pulled and serve the `Ci3_docker` module, which, btw, if you bother to scroll up 130 lines is just the default" :D

Reply:
`Current_docker.Default` means that Current_docker uses the local Docker context.  The value appears twice as this code is deployed to two different machines.  Each of those machines uses its local context.  So `Ci3_docker` and `Deploycamlorg_docker` are both the local Docker context but relative to the machines they are deployed on.  Changing both to read `Current_docker.Default` may obscure the fact they are on different machines.